packages/store-ui/src/organisms/Carousel/hooks/useCarousel.ts

```diff
@@ -0,0 +1,172 @@
+   */
+  currentSlide: number
+  /** Current active page */
+  currentPage: number
+  /** Whether or not the Carousel is currently sliding */
+  sliding: boolean
+  slideDirection: SlideDirection
+  totalItems: number
+  itemsPerPage: number
+  totalPages: number
+}
+
+export interface UseCarouselArgs {
+  totalItems: number
+  itemsPerPage?: number
+  swipeableConfigOverrides?: SwipeableProps
+}
+
+function reducer(state: CarouselState, action: Action): CarouselState {
+  switch (action.type) {
+    case 'NEXT_PAGE': {
+      let nextSlide = state.currentSlide + state.itemsPerPage
+      let nextPage = state.currentPage + 1
+
+      if (nextSlide > state.totalItems - 1) {
+        nextSlide = 0
+      }
+
+      if (nextPage >= state.totalPages) {
+        nextPage = 0
+      }
+
+      return {
+        ...state,
+        sliding: true,
+        slideDirection: 'next',
+        currentSlide: nextSlide,
+        currentPage: nextPage,
+      }
+    }
+
+    case 'PREVIOUS_PAGE': {
+      let previousSlide = state.currentSlide - state.itemsPerPage
+      let previousPage = state.currentPage - 1
+
+      if (previousSlide < 0) {
+        previousSlide = state.totalItems - state.itemsPerPage + 1
+      }
+
+      if (previousPage < 0) {
+        previousPage = state.totalPages - 1
+      }
+
+      return {
+        ...state,
+        sliding: true,
+        slideDirection: 'previous',
+        currentSlide: previousSlide,
+        currentPage: previousPage,
+      }
+    }
+
+    case 'GO_TO_PAGE': {
+      if (action.payload.pageIndex === state.currentPage) {
+        return state
+      }
+
+      return {
+        ...state,
+        sliding: true,
+        slideDirection:
+          action.payload.pageIndex > state.currentPage ? 'next' : 'previous',
+        currentSlide: action.payload.pageIndex * state.itemsPerPage,
+        currentPage: action.payload.pageIndex,
+      }
+    }
+
+    case 'STOP_SLIDE':
+      return { ...state, sliding: false }
+
+    default:
+      return state
+  }
+}
+
+const defaultCarouselState = (
+  totalItems: number,
+  itemsPerPage: number
+): CarouselState => ({
+  currentSlide: 0,
+  currentPage: 0,
+  sliding: false,
+  slideDirection: 'next',
+  totalItems,
+  itemsPerPage,
+  totalPages: Math.ceil(totalItems / itemsPerPage),
+})
+
+const slide = (slideDirection: SlideDirection, dispatch: Dispatch<Action>) => {
+  if (slideDirection === 'next') {
+    dispatch({
+      type: 'NEXT_PAGE',
+    })
+  }
+
+  if (slideDirection === 'previous') {
+    dispatch({
+      type: 'PREVIOUS_PAGE',
+    })
+  }
+
+  setTimeout(() => {
+    dispatch({ type: 'STOP_SLIDE' })
+  }, 50)
+}
+
+export const useCarousel = ({
+  totalItems,
+  swipeableConfigOverrides,
+  itemsPerPage = 1,
+}: UseCarouselArgs) => {
+  const [carouselState, carouselDispatch] = useReducer(
+    reducer,
+    defaultCarouselState(totalItems, itemsPerPage)
+  )
+
+  const handlers = useSwipeable({
```

Comment:
Do you know if will cause unnecessary rerender? If yes, you can put it into useMemo

Reply:
I don't see why it would cause unnecessary rerenders... could you give me an example?

Reply:
Nope, it was a mistake, this hook returns a memo value.